### ISSUE TITLE
Implement default using default

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#103]
On StackOverflow, [someone suggested](http://stackoverflow.com/questions/30949498/implementing-default-on-nested-structures-causes-a-stack-overflow) the compiler should keep them from implementing default in terms of itself.

Well, it's not exactly a common use case, but it's also pretty easy to spot.

**Comment by llogiq** [comment]
It just occurred to me that this is a special case of a more general 'unbounded_recursion' lint, only for trait impls.

**Comment by mcarton** [comment]
This can be closed, Rust has a `unconditional_recursion` lint since v1.3 (2015-09-17) which warns exactly for that:
```rust
% cat a.rs 
use std::default;

pub struct A {
    x: i32
}

impl Default for A {
    fn default() -> A {
        A { ..Default::default() }
    }
}

fn main() {
    let test = A { ..Default::default() };
}

% rustc a.rs 
[…]
a.rs:8:5: 10:6 warning: function cannot return without recurring, #[warn(unconditional_recursion)] on by default
a.rs: 8     fn default() -> A {
a.rs: 9         A { ..Default::default() }
a.rs:10     }
a.rs:9:15: 9:33 note: recursive call site
a.rs:9         A { ..Default::default() }
                     ^~~~~~~~~~~~~~~~~~
a.rs:8:5: 10:6 help: a `loop` may express intention better if this is on purpose
[…]
                ^~~~
```
and there also is #428 for a more general recursion lint.
